package request: FreeFileSync

**Project description**

> FreeFileSync is a folder comparison and synchronization software that creates and manages backup copies of all your important files. Instead of copying every file every time, FreeFileSync determines the differences between a source and a target folder and transfers only the minimum amount of data needed. FreeFileSync is Open Source software, available for Windows, macOS, and Linux.

This was the only open source tool I could find that does hassle-free *continuous* two-way sync with google drive on linux.

**Metadata**

* homepage URL: https://freefilesync.org/
* source URL: https://freefilesync.org/download/FreeFileSync_11.25_Source.zip (obtained from https://freefilesync.org/download.php)
  Comment: odd source distribution model; maybe the authors don't use version control software. 
* license: GPL Version 3 (see `License.txt` in source)
* platforms: windows, macos, linux

```
g++ -std=c++2b -pipe -DWXINTL_NO_GETTEXT_MACRO -I../.. -I../../zenXml -include "zen/i18n.h" -include "zen/warn_static.h" -Wall -Wfatal-errors -Wmissing-include-dirs -Wswitch-enum -Wcast-align -Wnon-virtual-dtor -Wno-unused-function -Wshadow -Wno-maybe-uninitialized -O3 -DNDEBUG `wx-config --cxxflags --debug=no` -pthread `pkg-config --cflags openssl` `pkg-config --cflags libcurl` `pkg-config --cflags libssh2` `pkg-config --cflags gtk+-2.0` -isystem/usr/include/gtk-2.0 -c application.cpp -o /build/FreeFileSync__Make/ffs/src/application.cpp.o
cc1plus: warning: /usr/include/gtk-2.0: No such file or directory [-Wmissing-include-dirs]
In file included from ui/file_grid_attr.h:12,
                 from config.h:14,
                 from application.h:13,
                 from application.cpp:7:
../../wx+/dc.h:15:10: fatal error: wx/bmpbndl.h: No such file or directory
   15 | #include <wx/bmpbndl.h>
      |          ^~~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:124: /build/FreeFileSync__Make/ffs/src/application.cpp.o] Error 1
```

Note that bmpbndl.h is available since wxGTK [3.1.6](https://docs.wxwidgets.org/trunk/bmpbndl_8h.html), while we only have 3.1.5 in nixpkgs. Ping #190946.

--- a/pkgs/applications/networking/freefilesync/default.nix
+++ b/pkgs/applications/networking/freefilesync/default.nix
@@ -1,0 +1,103 @@
+{ lib
+, gcc12Stdenv
+, fetchFromGitHub
+, fetchpatch
+, pkg-config
+, curl
+, glib
+, gtk3
+, libssh2
+, openssl
+, wxGTK32
+}:
+
+gcc12Stdenv.mkDerivation rec {
+  pname = "freefilesync";
+  version = "11.25";
+
+  src = fetchFromGitHub {
+    owner = "hkneptune";
+    repo = "FreeFileSync";
+    rev = "v${version}";
+    sha256 = "sha256-JV9qwBiF9kl+wc9+7lUufQVu6uiMQ6vojntxduNJ8MI=";
+  };
+
+  # Patches from ROSA Linux
+  patches = [
+    # Disable loading of the missing Animal.dat
+    (fetchpatch {
+      url = "https://abf.io/import/freefilesync/raw/rosa2021.1-11.25-1/ffs_devuan.patch";
+      sha256 = "sha256-o8T/tBinlhM1I82yXxm0ogZcZf+uri95vTJrca5mcqs=";
+      excludes = [ "FreeFileSync/Source/ffs_paths.cpp" ];
+      postFetch = ''
+        substituteInPlace $out --replace " for Rosa" ""
+      '';
+    })
+    # Fix build with GTK 3
+    (fetchpatch {
+      url = "https://abf.io/import/freefilesync/raw/rosa2021.1-11.25-1/ffs_devuan_gtk3.patch";
+      sha256 = "sha256-NXt/+BRTcMk8bnjR9Hipv1NzV9YqRJqy0e3RMInoWsA=";
+      postFetch = ''
+        substituteInPlace $out --replace "-isystem/usr/include/gtk-3.0" ""
+      '';
+    })
+  ];
+
+  postPatch = ''
+    substituteInPlace FreeFileSync/Source/ui/version_check.cpp \
+      --replace "openBrowserForDownload();" "openBrowserForDownload(parent);"
+  '';
+
+  nativeBuildInputs = [
+    pkg-config
+  ];
+
+  buildInputs = [
+    curl
+    glib
+    gtk3
+    libssh2
+    openssl
+    wxGTK32
+  ];
+
+  NIX_CFLAGS_COMPILE = [
+    # Undef g_object_ref on GLib 2.56+
+    "-DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_54"
+    "-DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_54"
+    # Define libssh2 constants
+    "-DMAX_SFTP_READ_SIZE=30000"
+    "-DMAX_SFTP_OUTGOING_SIZE=30000"
+  ];
+
+  buildPhase = ''
+    runHook preBuild
+
+    chmod +w FreeFileSync/Build
+    cd FreeFileSync/Source
+    make -j$NIX_BUILD_CORES
+    cd RealTimeSync
+    make -j$NIX_BUILD_CORES
+    cd ../../..
+
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out
+    cp -R FreeFileSync/Build/* $out
+    mv $out/{Bin,bin}
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "Open Source File Synchronization & Backup Software";
+    homepage = "https://freefilesync.org";
+    license = licenses.gpl3Only;
+    maintainers = with maintainers; [ wegank ];
+    platforms = platforms.linux;
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -27859,6 +27859,8 @@ with pkgs;
 
   freebayes = callPackage ../applications/science/biology/freebayes { };
 
+  freefilesync = callPackage ../applications/networking/freefilesync { };
+
   freewheeling = callPackage ../applications/audio/freewheeling { };
 
   fritzing = libsForQt5.callPackage ../applications/science/electronics/fritzing { };

